new module: aix_devices, manage AIX devices

kairoaraujo opened the pull request:
##### SUMMARY
New module for AIX.
This module discovers, defines, removes and modifies attributes of AIX devices.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/system/aix_devices.py

##### ANSIBLE VERSION
```
ansible 2.5.0 (aix_devices 975f695d1b) last updated 2017/10/28 12:39:51 (GMT +200)
  config file = None
  configured module search path = ['/Users/kairo/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/kairo/Dev/Ansible/modules/aix_devices/ansible/lib/ansible
  executable location = /Users/kairo/Dev/Ansible/modules/aix_devices/ansible/bin/ansible
  python version = 3.6.3 (v3.6.3:2c5fed86e0, Oct  3 2017, 00:32:08) [GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]
```


##### ADDITIONAL INFORMATION
There is an example of playbook testing possibilities and playbook returns

```yml
---
- name: Testing
  hosts: lpar22

  tasks:
    - name: Scan new devices.
      aix_devices:
        device: all
        state: present

    - name: Scan new virtual devices (vio0).
      aix_devices:
        device: vio0
        state: present

    - name: Removes ent2.
      aix_devices:
        device: ent2
        state: absent

    - name: Put device en2 in Defined
      aix_devices:
        device: en2
        state: defined

    - name: Removes ent4 (inexistent).
      aix_devices:
        device: ent4
        state: absent

    - name: Put device en4 in Defined (inexistent)
      aix_devices:
        device: en4
        state: defined

    - name: Put vscsi1 and children devices in Defined state.
      aix_devices:
        device: vscsi1
        recursive: yes
        state: defined

    - name: Removes vscsi1 and children devices.
      aix_devices:
        device: vscsi1
        recursive: yes
        state: absent

    - name: Changes en1 mtu to 9000 and disables arp.
      aix_devices:
        device: en1
        attributes: mtu=900,arp=off
        state: present

    - name: Configure IP, netmask and set en1 up.
      aix_devices:
        device: en1
        attributes: netaddr=192.168.0.100,netmask=255.255.255.0,state=up
        state: present

    - name: Adding IP alias to en0
      aix_devices:
        device: en0
        attributes: [
            "alias4=10.0.0.100,255.255.255.0"
            ]
        state: present
```

Playbook running:
```console
ansible-playbook -i ../../../inventory ../test.yml -v
No config file found; using defaults

PLAY [Testing] ************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************
ok: [lpar22]

TASK [Scan new devices.] **************************************************************************************
changed: [lpar22] => {"changed": true, "failed": false, "msg": "", "state": "present"}

TASK [Scan new virtual devices (vio0).] ***********************************************************************
changed: [lpar22] => {"changed": true, "failed": false, "msg": "", "state": "present"}

TASK [Removes ent2.] ******************************************************************************************
changed: [lpar22] => {"changed": true, "failed": false, "msg": "ent2 deleted\n", "state": "absent"}

TASK [Put device en2 in Defined] ******************************************************************************
ok: [lpar22] => {"changed": false, "failed": false, "msg": "Device en2 already in Defined", "state": "defined"}

TASK [Removes ent4 (inexistent).] *****************************************************************************
ok: [lpar22] => {"changed": false, "failed": false, "msg": "Device ent4 does not exist.", "state": "absent"}

TASK [Put device en4 in Defined (inexistent)] *****************************************************************
ok: [lpar22] => {"changed": false, "failed": false, "msg": "Device en4 does not exist.", "state": "defined"}

TASK [Put vscsi1 and children devices in Defined state.] ******************************************************
changed: [lpar22] => {"changed": true, "failed": false, "msg": "vscsi1 Defined\n", "state": "defined"}

TASK [Removes vscsi1 and children devices.] *******************************************************************
changed: [lpar22] => {"changed": true, "failed": false, "msg": "vscsi1 deleted\n", "state": "absent"}

TASK [Changes en1 mtu to 9000 and disables arp.] **************************************************************
changed: [lpar22] => {"changed": true, "failed": false, "msg": "Attributes changed: arp=off. Attributes already set: mtu=900. ", "state": "present"}

TASK [Configure IP, netmask and set en1 up.] ******************************************************************
changed: [lpar22] => {"changed": true, "failed": false, "msg": "Attributes changed: netaddr=192.168.0.100,netmask=255.255.255.0,state=up. ", "state": "present"}

TASK [Adding IP alias to en0] *********************************************************************************
changed: [lpar22] => {"changed": true, "failed": false, "msg": "Attributes changed: alias4=10.0.0.100,255.255.255.0. ", "state": "present"}

PLAY RECAP ****************************************************************************************************
lpar22                     : ok=12   changed=8    unreachable=0    failed=0



```

kairoaraujo committed:
new module: aix_devices AIX devices management

This module discovery, defines, removes and modifies attributes of AIX
devices.

kairoaraujo committed:
Added hide attributes that can be used for aliases

Added hid attributes that can be used to manage aliases on en
interfaces.

kairoaraujo committed:
After tests: docs and attributes tests

Fixed attributes tests and doc explaining how to use
attributes with comma.

kairoaraujo committed:
Fixed grammar on module description

Fixed grammar on module description

kairoaraujo committed:
Included test/legacy/aix_devices.yml for tests

As discussed on IRC ansible-devel channes, was include the
legacy tests for further manual tests.

alikins commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,382 @@
+    description:
+      - Specifies attributes to device separated by comma.
+      - If attribute has comma, explicit list using
+        [ "parameter=value,with,comma", "parameter=foobar" ]. See alias
```
Could attributes be a dict here?

Might eliminate the need for the comma splitting logic.


``` yaml
aix_devices:
  device: en1
  attributes:
    mtu: 900
    arp: off
    some_other_attr: blip
    alias4: 10.0.0.100,255.255.255.0
  state: present

```

alikins commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,382 @@
+
+    """
+    lsattr_cmd = module.get_bin_path("lsattr", True)
+    rc, lsattr_out, err = module.run_command("%s -El %s -a %s" % (lsattr_cmd, device, attr))
```
If the attribute values can contain commas or whitespace, using run_command with cmd args as a list may
avoid problems here.

alikins commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,382 @@
+        return current_param
+
+    elif rc != 0:
+        module.fail_json(msg="Failed to run lsattr.", rc=rc, err=err)
```
If the lsattr args aren't sensitive I would include the full cmd in the error message here.

Commit:
Added 'attributes' as dictionary

Added 'attributes' as a dictionary makes the configuration
simple.

Commit:
Changed the added version from 2.5 to 2.7

Fixed the shippable error from 2.5 to 2.7

```
2018-06-01 08:28:02 ERROR: Found 1 validate-modules issue(s) which
need to be resolved:
2018-06-01 08:28:02 ERROR:
lib/ansible/modules/system/aix_devices.py:0:0: E307 version_added
should be 2.7. Currently 2.5 (75%)
```

kairoaraujo commented:
bot_status

maxamillion commented:
cc @alikins

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+author:
+  - Kairo Araujo (@kairoaraujo)
+module: aix_devices
+short_description: Manages AIX devices.
```
```suggestion
short_description: Manages AIX devices
```
The short description has no trailing dot, all other descriptions do.

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+description:
+  - This module discovers, defines, removes and modifies attributes of AIX
+    devices.
+version_added: "2.7"
```
```suggestion
version_added: '2.8'
```
Update to new release.

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+options:
+  attributes:
+    description:
+      - Specifies attributes to device separated by comma.
```
```suggestion
      - Specify attributes for this device.
      - The value is a dictionary of attributes with values.
    type: dict
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+      - Specifies attributes to device separated by comma.
+  device:
+    description:
+      - Device name.
```
```suggestion
      - The name of the device.
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+  device:
+    description:
+      - Device name.
+      - C(all) is valid to rescan C(present) all devices (AIX cfgmgr command).
```
```suggestion
      - Use C(all) to rescan all C(present) devices (AIX cfgmgr command).
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+    description:
+      - Device name.
+      - C(all) is valid to rescan C(present) all devices (AIX cfgmgr command).
+    required: true
```
```suggestion
    required: true
    type: str
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+    required: true
+  force:
+    description:
+      - Forces action.
```
We tend to prefer a more descriptive option rather than this generic *force*.
So what does force actually do ?

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+    description:
+      - Forces action.
+    type: bool
+    default: "no"
```
```suggestion
    default: no
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+    default: "no"
+  recursive:
+    description:
+      - Removes or defines a device and children devices.
```
```suggestion
      - Whether to remove or define a device including child devices.
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+    description:
+      - Removes or defines a device and children devices.
+    type: bool
+    default: "no"
```
```suggestion
    default: no
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+        not specified).
+      - C(absent) removes a device.
+      - C(defined) changes device to Defined state.
+    choices: [present, absent, defined]
```
```suggestion
    required: true
    type: str
    choices: [ absent, defined, present ]
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+      - C(defined) changes device to Defined state.
+    choices: [present, absent, defined]
+    default: present
+    required: true
```
```suggestion
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+      - C(present) rescan a specific device or all devices (when C(device) is
+        not specified).
+      - C(absent) removes a device.
+      - C(defined) changes device to Defined state.
```
I think this is confusing, `present` means rescan?
What does "defined state" means? And are they mutually exclusive ?

gforster replied:
This has to do with the ODM in AIX which is a database for the device information. You can have something defined that is not available. Maybe we should have in the description how this is working with ODM?

dagwieers replied:
Guess so, if this make sense for people that would use this, it is fine by me though.

kairoaraujo replied:
`present` corresponds to `available` in AIX, that is the normal state (Should we have an alias?).
`defined` is the state used by AIX sysadmins to put the device in maintenance mode, for example to be replaced an Ethernet card or disk, but the informations will be kept in the AIX ODM.

gforster replied:
I like the alias idea. May make sense to most AIX admins I would think.

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+  description: Return changed for aix_device actions as true or false.
+  returned: always
+  type: boolean
+  version_added: "2.7"
```
```suggestion
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+  description: Return message regarding the action.
+  returned: always
+  type: string
+  version_added: "2.7"
```
```suggestion
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+            else:
+                result['msg'] = "Device %s does not exist." % device
+
+            module.exit_json(**result)
```
```suggestion
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+            else:
+                result['changed'], result['msg'] = discover_device(module, device)
+
+            module.exit_json(**result)
```
```suggestion
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+        if not device:
+            result['msg'] = "device is required to absent or defined state."
+
+            module.exit_json(**result)
```
```suggestion
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+            else:
+                result['msg'] = "Device %s does not exist." % device
+
+            module.exit_json(**result)
```
```suggestion
```

dagwieers reviewed (changes requested):
LGTM, just a few cosmetic/docs changes.

And the tests should not go into test/legacy, but in test/integration/targets (with aliases containing `unsupported`).

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+changed:
+  description: Return changed for aix_device actions as true or false.
+  returned: always
+  type: boolean
```
```suggestion
  type: bool
```

dagwieers commented on lib/ansible/modules/system/aix_devices.py:
```diff
@@ -0,0 +1,406 @@
+msg:
+  description: Return message regarding the action.
+  returned: always
+  type: string
```
```suggestion
  type: str
```

dagwieers committed:
Various changes

dagwieers committed:
Revert

kairoaraujo committed:
Changed the tests to integration dir

kairoaraujo committed:
Implement 'available' state

'available' state is the AIX state used, that works same as 'present'

kairoaraujo committed:
The states were changed to AIX expressions and kept Ansible states.

Makes sense keep the states names to AIX and use the Ansible 'standards'
states.

'available' is 'present'
'removed' is 'absent'

It makes easy to AIX sysadmins use the module, however it keep the
Ansible meanings.

kairoaraujo committed:
Fixed choices according with latest patchset (commit)

gforster commented:
Looks good. I should be able to test tomorrow 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

kairoaraujo commented:
@ramooncamacho it is ready for tests

dagwieers committed:
A few doc changes

Nothing material

dagwieers commented:
rebuild_merge

ramooncamacho commented:
@dagwieers and @kairoaraujo, should I just comment my tests or do I open the issues? I will start sending the comments here:

**Issue 1:**

Please change in the example page from present to available, and from absent to removed, as the options have changed.

**Issue 2:**

Trying to discover all devices, it was returned changed as true, but the new device didn't show up. 

I think it should be expected this command be always status changed, but I noticed an hidden error, because in AIX if an object class doesn't exist, the return code is still 0:

Trying to discover all devices:

```
[testuser@controller ansible]$ ansible -s -m aix_devices -a 'device=all state=available' testeram
testeram | CHANGED => {
    "changed": true,
    "msg": ""
}
[testuser@controller ansible]$ 
```

I realized the new devices didn't show up:

```
testuser@testeram(/home/testuser)# lsdev -C | grep fcs
testuser@testeram(/home/testuser)#
```

The module is trying to execute "cfgmgr -l all", and doing it manually, the command is trying to find a device named "all". To discover all devices I have always used cfgmgr with no option, as below:

```
testuser@testeram(/home/testuser)# lsdev -C | grep fcs
testuser@testeram(/home/testuser)# sudo cfgmgr -l all
cfgmgr: 0514-604 Cannot access the CuDv object class in the
        device configuration database.
testuser@testeram(/home/testuser)# lsdev -C | grep fcs
testuser@testeram(/home/testuser)# sudo cfgmgr
testuser@testeram(/home/testuser)# lsdev -C | grep fcs
fcs0       Available 29-T1       Virtual Fibre Channel Client Adapter
testuser@testeram(/home/testuser)#
```

Return code in the condition the object class is not found:

```
testuser@testeram(/home/testuser)# sudo cfgmgr -l all
cfgmgr: 0514-604 Cannot access the CuDv object class in the
        device configuration database.
testuser@testeram(/home/testuser)# echo $?
0
testuser@testeram(/home/testuser)#
```

**Issue 3:**

Trying to remove an adapter, it returns error:

```
[testuser@controller ansible]$ cat aix_dev.yaml
---
- name: Testing
  hosts: testeram
  become: yes

  tasks:
    - name: Removes fcs0.
      aix_devices:
        device: fcs0
        state: removed
        recursive: yes
[testuser@controller ansible]$
```

```
[testuser@controller ansible]$ ansible-playbook aix_dev.yaml

PLAY [Testing] *********************************************************************************************************************************************************

TASK [Gathering Facts] *************************************************************************************************************************************************
ok: [testeram]

TASK [Removes fcs0.] ***************************************************************************************************************************************************
fatal: [testeram]: FAILED! => {"changed": false, "err": "rmdev: 0514-500 Usage error - \n\nUsage:\nrmdev {-l|-p} Name [-d|-S][-R][-q][-f File][-g]\nrmdev -h\n\n", "msg": "Failed to run rmdev", "rc": 1}
        to retry, use: --limit @/home/testuser/ansible/aix_dev.retry

PLAY RECAP *************************************************************************************************************************************************************
testeram                   : ok=1    changed=0    unreachable=0    failed=1    skipped=0

[testuser@controller ansible]$
```

I saw it's trying to execute "/usr/sbin/rmdev -l fcs0 -R ''", and doing the same at the operating system, it returns error:

```
testuser@testeram(/home/testuser)# sudo /usr/sbin/rmdev -l fcs0 -R ''
rmdev: 0514-500 Usage error -

Usage:
rmdev {-l|-p} Name [-d|-S][-R][-q][-f File][-g]
rmdev -h

testuser@testeram(/home/testuser)#
```

If I do the same without the single quotation marks, it executes correctly:

```
testuser@testeram(/home/testuser)# sudo /usr/sbin/rmdev -l fcs0 -R
sfwcomm0 Defined
fscsi0 Defined
fcs0 Defined
testuser@testeram(/home/testuser)#
```

dagwieers commented:
@ramooncamacho Please open issues for each. Thanks for testing !

ramooncamacho commented:
The issues were opened:

https://github.com/ansible/ansible/issues/51750
https://github.com/ansible/ansible/issues/51754
https://github.com/ansible/ansible/issues/51755
